dashboard/src/actions/authActions.js

```diff
@@ -7,6 +7,49 @@ import Cookies from "js-cookie";
 import { showToast } from "actions/toastActions";
 import { uid } from "../utils/helper";
 
+
+// Create an Authentication Request
+export const authenticationRequest = () => async (dispatch, getState) => {
+    try {
+      const endpoints = getState().apiEndpoint.endpoints;
+      const oidcServer = endpoints["openid-connect"]?.issuer;
+      const oidcRealm = endpoints["openid-connect"]?.realm;
+      const oidcClient = endpoints["openid-connect"]?.client;
+      const oidcClientSecret = endpoints["openid-connect"]?.secret;
+      let req = oidcServer + '/realms/' + oidcRealm + '/protocol/openid-connect/auth';
+      req += '?client_id=' + oidcClient;
+      req += '&client_secret=' + oidcClientSecret;
```

Comment:
This doesn't feel right.  Maybe I really just don't understand OpenID Connect, but why would the server give its client secret to a browser to use in a request?

Reply:
Okay, so Pbench Server is providing the Dashboard a set of client credentials to use to access the Keycloak broker. Without these credentials, the Dashboard will have to access the OIDC via the Pbench server and we don't want that we want the dashboard to handle all the redirects on its own. Now we can have multiple `clients` under our OIDC `Realm` For example I can list the following scenarios:

1st:
```
Pbench Realm

1. pbench-server client  # Private client with a secret that should not be passed around.
2. pbench-dashoard client # Public client with no secret

```
2nd:
```
Pbench Realm

1. pbench-server client  # Private client with a secret that should not be passed around.
2. pbench-dashoard client # Private client with a secret that can be shared publicly (acts as a public client with a secret)

```
3rd:
```
Pbench Realm

1. pbench-server client  # Private client with a secret that can be shared publicly
```
In the first scenario, Pbench Server will hand out the `pbench-dashoard` client credentials without any secret since there is none. The dashboard will use these credentials to make a request to the OIDC server, If the OIDC server approves the request and the user performs a successful login then the dashboard will get back the user tokens from the OIDC server.
The 2nd and 3rd scenarios are the exact same but the request will have a client's secret (which is open). The dashboard will use these tokens to make API requests on the Pbench server where server will decode these tokens without communicating with OIDC server (offline). If the user passes all the required checks they will be granted access to the resource.

I started all the Keycloak and OpenID efforts without making things complicated where we have 1 realm and 1 public client with a secret. However, we could do it either way; if we want to create a separate client in our Keycloak configuration, we could. And we certainly will for the staging and production but this is configuration stuff we need to do in the Pbench Server and it has no impact on the dashboard. All dashboard does is that it takes the credentials from the server and use it to make a request the OIDC server. It's up to the server to decide which credentials it will share.

Does this make sense?

Reply:
Probably what I should do is add a conditional statement to see if the `oidcClientSecret` is present in the response payload of the endpoints API. if not present then we would not add this uri parameter in the request otherwise add it if the server gave one.

Reply:
Based on https://www.rfc-editor.org/rfc/rfc6749#section-1.1:

 * "resource owner" the end-user sitting at keyboard
 * "resource server" the pbench server
 * "client" the pbench dashboard code running in a browser
 * "authorization server" the OpenID Connect identity provider / broker

The "Client ID" is for the dashboard not the server.  The Pbench Server hands the dashboard the "Client ID" it must use to talk to the authorization server, but the pbench server itself is not a client because it is not accessing resources on a resource server, that is what the dashboard is doing.

So to unravel this, where is the documentation for the KeyCloak API you reference above, `oidcServer + '/realms/' + oidcRealm + '/protocol/openid-connect/auth'`?

Reply:
https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth and https://www.rfc-editor.org/rfc/rfc6749#section-4.1.1

Reply:
Also [this](https://github.com/keycloak/keycloak-documentation/blob/main/securing_apps/topics/oidc/oidc-generic.adoc) is a Keycloak github doc page where you will see the authorization endpoint and pair that with https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationEndpoint to see how the request parameters look like.